tiled/client/base.py

```diff
@@ -364,11 +374,139 @@ def formats(self):
         )
         return sorted(formats)
 
-    def update_metadata(self, metadata=None, specs=None):
+    def update_metadata(self, metadata=None, specs=None, exec=True):
+        """
+        EXPERIMENTAL: Update metadata via a `dict.update`- like interface.
+
+        `update_metadata` is a user-friendly wrapper for `patch_metadata`.
+        This is subject to change or removal without notice.
+
+        Parameters
+        ----------
+        metadata : dict, optional
+            User metadata. May be nested. Must contain only basic types
+            (e.g. numbers, strings, lists, dicts) that are JSON-serializable.
+        specs : List[str], optional
+            List of names that are used to label that the data and/or metadata
+            conform to some named standard specification.
+        exec : boolean
+            Defaulots to True. If set to False, ignores specs and returns
+            a JSON patch for metadata without applying it.
+
+        See Also
+        --------
+        patch_metadata
+        replace_metadata
+
+        Notes
+        -----
+        `update_metadata` constructs a JSON Patch (RFC6902) by comparing user updates
+        to existing metadata. It uses a slight variation of JSON Merge Patch (RFC7386)
+        as an intermediary to implement a python `dict.update`-like user-friendly
+        interface, but with additional features like key deletion (see examples) and
+        support for `None (null)` values.
+
+        Examples
+        --------
+
+        Add or update a key-value pair at the top or a nested level
+
+        >>> node.update_metadata({'key': new_value})
+        >>> node.update_metadata({'top_key': {'nested_key': new_value}})
+
+        Remove an existing key
+
+        >>> from tiled.client.metadata_update import DELETE_KEY
+        >>> node.update_metadata({'key_to_be_deleted': DELETE_KEY})
+
+        Interactively update complex metadata using a copy of original structure
+        (e.g., in iPython you may use tab completion to navigate nested metadata)
+
+        >>> md = node.metadata_copy()
+        >>> md['L1_key']['L2_key']['L3_key'] = new_value  # use tab completion
+        >>> md['unwanted_key'] = DELETE_KEY
+        >>> node.update_metadata(metadata=md)  # Update the copy on the server
+        """
+
+        if metadata is None:
+            metadata = {}
+
+        patch = jsonpatch.JsonPatch.from_diff(
+            dict(self.metadata),
+            apply_update_patch(self.metadata_copy(), metadata),
+            dumps=orjson.dumps,
+        ).patch
+
+        if not exec:
```

Comment:
I like the addition of a way to inspect the patch before applying it. I'd like to explore other ways to implement that, because this violates the principle that we generally try to follow: an optional parameter should not alter the return type.

I can think of two ways to achieve this:

1. Pull this out into a convenience function, `build_patch`.
2. Make it so that `update_metadata` returns the patch unconditionally, and have `exec` (or perhaps the inverse, `dry_run`) control whether or not it is applied, not what is returned. This might cut against the "user friendly" purpose though, as the user looking at the return value might wonder, "What am I supposed to _do_ with this? Did it work?" Perhaps this points toward (1).

Reply:
Thanks @danielballan, as is always this is another very helpful comment. I "knew" I was violating some design principle, but couldn't figure which one. I am leaning toward the convenience function approach for the same reason you mention in 2.